s3/replication/common/src/s3replicationcommon/s3_head_object.py

```diff
@@ -0,0 +1,122 @@
+#
+# Copyright (c) 2021 Seagate Technology LLC and/or its Affiliates
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+from s3replicationcommon.s3_common import S3RequestState
+from s3replicationcommon.timer import Timer
+from s3replicationcommon.aws_v4_signer import AWSV4Signer
+from s3replicationcommon.log import fmt_reqid_log
+import sys
+import aiohttp
+
+
+class S3AsyncHeadObject:
+    def __init__(self, session, request_id,
+                 bucket_name, object_name):
+        # initialize
+        self._session = session
+        # Request id for better logging.
+        self._request_id = request_id
+        self._logger = session.logger
+
+        self._bucket_name = bucket_name
+        self._object_name = object_name
+
+        self.remote_down = False
+        self._http_status = None
+
+        self._timer = Timer()
+        self._state = S3RequestState.INITIALISED
+
+    async def get_head_object(self):
+        request_uri = AWSV4Signer.fmt_s3_request_uri(
+            self._bucket_name, self._object_name)
+
+        query_params = ""
+        body = ""
+        headers = AWSV4Signer(
+            self._session.endpoint,
+            self._session.service_name,
+            self._session.region,
+            self._session.access_key,
+            self._session.secret_key).prepare_signed_header(
+            'HEAD',
+            request_uri,
+            query_params,
+            body
+        )
+
+        if (headers['Authorization'] is None):
+            self._logger.error(fmt_reqid_log(self._request_id) +
+                               "Failed to generate v4 signature")
+            sys.exit(-1)
+
+        self._logger.info(fmt_reqid_log(self._request_id) +
+                          'HEAD on {}'.format(
+                              self._session.endpoint + request_uri))
+        self._logger.debug(fmt_reqid_log(self._request_id) +
+                           "HEAD Request Header {}".format(headers))
+
+        self._timer.start()
+        try:
+            async with self._session.get_client_session().head(
+                    self._session.endpoint + request_uri,
+                    headers=headers) as resp:
+                if resp.status == 200:
```

Comment:
in success case, remember returned headers which is the main purpose of this API.
`self._response_headers = resp.headers`

Reply:
we will need getters to read the received headers. Check response syntax here https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html

Reply:
done